File: solutions/typescript/luhn/luhn.ts
/**
 * Given a number determine whether or not it is valid per the Luhn formula.
 *
 * @param {String} input
 * @returns {Boolean}
 */
export function valid(input: String): Boolean {
  const digits = input.replace(/\s/g, "");
  if (digits.length < 2 || digits.match(/\D/)) return false;

  let sum = 0;
  let odd = true;
  for (const d of digits.split("").reverse().map(Number)) {
    sum += odd ? d : d < 5 ? d * 2 : d * 2 - 9;
    odd = !odd;
  }
  return sum % 10 == 0;
}

File: solutions/ruby/luhn/luhn.rb
class Luhn
  def self.valid?(input)
    input
      .gsub(/\s/, "")
      .tap { |s| return false unless s[/^\d{2,}$/] }
      .chars
      .reverse
      .map.with_index { |d, i| i.odd? ? d.to_i * 2 : d.to_i }
      .map { |d| d > 9 ? d - 9 : d }
      .sum % 10 == 0
  end
end
